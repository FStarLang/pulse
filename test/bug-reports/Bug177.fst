module Bug177

#lang-pulse
open Pulse

// Not great, but a common issue of trying to use a computation as a value
// - Expected type int
//   but !r
//   has type
//       stt int
//         (Pulse.Lib.Reference.pts_to r (*?u2491*) _)
//         (fun x ->
//             Pulse.Lib.Reference.pts_to r (*?u2491*) _ **
//             pure ((*?u2491*) _ == x))
[@@expect_failure]
fn kswap_int
  (r : ref int)
  requires pts_to r 'v
  ensures  pts_to r 'v
{
  r := !r;
}

// Much worse.
// - Ill-typed term: r := !r
// - Assertion failed
// - The SMT solver could not prove the query. Use --query_stats for more details.
[@@expect_failure]
fn kswap_t
  (#t : Type0)
  (r : ref t)
  requires pts_to r 'v
  ensures  pts_to r 'v
{
  r := !r;
}
