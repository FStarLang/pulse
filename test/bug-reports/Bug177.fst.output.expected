>> Got issues: [
* Error 12 at Bug177.fst(21,7-21,9):
  - Expected type Prims.int
    but !r
    has type
        Pulse.Lib.Core.stt Prims.int
          (Pulse.Lib.Reference.pts_to r (*?u247*) _)
          (fun x ->
              Pulse.Lib.Reference.pts_to r (*?u247*) _ **
              Pulse.Lib.Core.pure ((*?u247*) _ == x))

>>]
>> Got issues: [
* Error 19 at Bug177.fst(35,2-35,8):
  - Ill-typed term: r := !r
  - Assertion failed
  - The SMT solver could not prove the query. Use --query_stats for more
    details.
  - See also Bug177.fst(35,2-35,9)

>>]
